File: android/app/src/main/java/tech/loveace/appv3/ui/viewmodel/AcademicViewModel.kt
package tech.loveace.appv3.ui.viewmodel

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tech.loveace.appv3.data.model.*
import tech.loveace.appv3.data.service.JWCService

data class AcademicUiState(
    val isLoading: Boolean = false,
    val academicInfo: AcademicInfo? = null,
    val terms: List<TermItem> = emptyList(),
    val selectedTerm: TermItem? = null,
    val scores: TermScoreResponse? = null,
    val scoresLoading: Boolean = false,
    val selectedScore: ScoreRecord? = null,
    val scoreDetail: ScoreDetail? = null,
    val scoreDetailLoading: Boolean = false,
    val scoreDetailError: String? = null,
    val error: String? = null,
)

class AcademicViewModel : ViewModel() {
    private var service: JWCService? = null

    private val _uiState = MutableStateFlow(AcademicUiState())
    val uiState: StateFlow<AcademicUiState> = _uiState.asStateFlow()

    fun init(service: JWCService) {
        this.service = service
    }

    fun loadAcademicInfo() {
        val svc = service ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            val result = svc.getAcademicInfo()
            _uiState.value = if (result.success) {
                _uiState.value.copy(isLoading = false, academicInfo = result.data)
            } else {
                _uiState.value.copy(isLoading = false, error = result.error)
            }
        }
    }

    fun loadTerms() {
        val svc = service ?: return
        viewModelScope.launch {
            val result = svc.getAllTerms()
            if (result.success && result.data != null) {
                val terms = result.data
                val current = terms.firstOrNull { it.isCurrent } ?: terms.firstOrNull()
                _uiState.value = _uiState.value.copy(terms = terms, selectedTerm = current)
                current?.let { loadScores(it.termCode) }
            }
        }
    }

    fun selectTerm(term: TermItem) {
        _uiState.value = _uiState.value.copy(
            selectedTerm = term,
            selectedScore = null,
            scoreDetail = null,
            scoreDetailError = null,
        )
        loadScores(term.termCode)
    }

    fun loadScores(termCode: String) {
        val svc = service ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(scoresLoading = true)
            val result = if (_uiState.value.selectedTerm?.isCurrent == true) {
                svc.getThisTermScores()
            } else {
                svc.getTermScore(termCode)
            }
            _uiState.value = if (result.success) {
                _uiState.value.copy(scoresLoading = false, scores = result.data)
            } else {
                _uiState.value.copy(scoresLoading = false, error = result.error)
            }
        }
    }

    fun loadScoreDetail(record: ScoreRecord) {
        val svc = service ?: return
        _uiState.value = _uiState.value.copy(
            selectedScore = record,
            scoreDetail = null,
            scoreDetailLoading = true,
            scoreDetailError = null,
        )
        viewModelScope.launch {
            val result = svc.getScoreDetail(record)
            _uiState.value = if (result.success) {
                _uiState.value.copy(scoreDetail = result.data, scoreDetailLoading = false)
            } else {
                _uiState.value.copy(scoreDetailLoading = false, scoreDetailError = result.error)
            }
        }
    }

    fun dismissScoreDetail() {
        _uiState.value = _uiState.value.copy(
            selectedScore = null,
            scoreDetail = null,
            scoreDetailLoading = false,
            scoreDetailError = null,
        )
    }

    companion object {
        private const val TAG = "AcademicViewModel"
    }
}

File: ios/loveaceios/ViewModel/AcademicViewModel.swift
import Foundation

struct TermGPAPoint: Identifiable {
    var id: String { termName }
    let termName: String
    let termCode: String
    let gpa: Double
    let courseCount: Int
    let totalCredits: Double
}

struct ScoreDistBucket: Identifiable {
    var id: String { label }
    let label: String
    let count: Int
    let color: String
}

@MainActor
@Observable
final class AcademicViewModel {
    var isLoading = false
    var academicInfo: AcademicInfo?
    var terms: [TermItem] = []
    var selectedTerm: TermItem?
    var scores: TermScoreResponse?
    var scoresLoading = false
    var selectedScore: ScoreRecord?
    var scoreDetail: ScoreDetail?
    var scoreDetailLoading = false
    var scoreDetailError: String?
    var error: String?

    var allTermGPA: [TermGPAPoint] = []
    var allScores: [ScoreRecord] = []
    var insightsLoading = false

    private var service: JWCService?

    func initialize(service: JWCService) { self.service = service }

    func loadAcademicInfo() {
        guard let svc = service else { return }
        Task {
            isLoading = true; error = nil
            let result = await svc.getAcademicInfo()
            isLoading = false
            if result.success { academicInfo = result.data } else { error = result.error }
        }
    }

    func loadTerms() {
        guard let svc = service else { return }
        Task {
            let result = await svc.getAllTerms()
            if result.success, let data = result.data {
                terms = data
                let current = data.first { $0.isCurrent } ?? data.first
                selectedTerm = current
                if let tc = current { loadScores(termCode: tc.termCode) }
            }
        }
    }

    func selectTerm(_ term: TermItem) {
        selectedTerm = term
        selectedScore = nil
        scoreDetail = nil
        scoreDetailError = nil
        loadScores(termCode: term.termCode)
    }

    func loadScores(termCode: String) {
        guard let svc = service else { return }
        Task {
            scoresLoading = true
            let result = selectedTerm?.isCurrent == true
                ? await svc.getThisTermScores()
                : await svc.getTermScore(termCode: termCode)
            scoresLoading = false
            if result.success { scores = result.data } else { error = result.error }
        }
    }

    func loadScoreDetail(_ record: ScoreRecord) {
        guard let svc = service else { return }
        selectedScore = record
        scoreDetail = nil
        scoreDetailLoading = true
        scoreDetailError = nil
        Task {
            let result = await svc.getScoreDetail(record: record)
            scoreDetailLoading = false
            if result.success { scoreDetail = result.data } else { scoreDetailError = result.error }
        }
    }

    func dismissScoreDetail() {
        selectedScore = nil
        scoreDetail = nil
        scoreDetailLoading = false
        scoreDetailError = nil
    }

    func loadInsightsData() {
        guard let svc = service, !insightsLoading else { return }
        Task {
            insightsLoading = true
            if terms.isEmpty {
                let result = await svc.getAllTerms()
                if result.success, let data = result.data { terms = data }
            }

            var points: [TermGPAPoint] = []
            var allRecords: [ScoreRecord] = []

            for term in terms.reversed() {
                let result = await svc.getTermScore(termCode: term.termCode)
                guard result.success, let data = result.data, !data.records.isEmpty else { continue }
                allRecords.append(contentsOf: data.records)

                var totalWeighted = 0.0
                var totalCredits = 0.0
                for r in data.records {
                    guard let score = Double(r.score), let credit = Double(r.credits), credit > 0 else { continue }
                    let gp: Double
                    if score >= 60 { gp = (score - 50) / 10.0 }
                    else { gp = 0.0 }
                    totalWeighted += gp * credit
                    totalCredits += credit
                }
                if totalCredits > 0 {
                    points.append(TermGPAPoint(
                        termName: term.termName, termCode: term.termCode,
                        gpa: totalWeighted / totalCredits,
                        courseCount: data.records.count, totalCredits: totalCredits
                    ))
                }
            }

            allTermGPA = points
            allScores = allRecords
            insightsLoading = false
        }
    }

    var scoreDistribution: [ScoreDistBucket] {
        var buckets = [
            "90+": 0, "80-89": 0, "70-79": 0, "60-69": 0, "<60": 0
        ]
        for r in allScores {
            guard let s = Double(r.score) else { continue }
            if s >= 90 { buckets["90+", default: 0] += 1 }
            else if s >= 80 { buckets["80-89", default: 0] += 1 }
            else if s >= 70 { buckets["70-79", default: 0] += 1 }
            else if s >= 60 { buckets["60-69", default: 0] += 1 }
            else { buckets["<60", default: 0] += 1 }
        }
        return [
            ScoreDistBucket(label: "90+", count: buckets["90+"]!, color: "green"),
            ScoreDistBucket(label: "80-89", count: buckets["80-89"]!, color: "blue"),
            ScoreDistBucket(label: "70-79", count: buckets["70-79"]!, color: "orange"),
            ScoreDistBucket(label: "60-69", count: buckets["60-69"]!, color: "yellow"),
            ScoreDistBucket(label: "<60", count: buckets["<60"]!, color: "red"),
        ]
    }

    var bestSubject: ScoreRecord? {
        allScores.compactMap { r -> (ScoreRecord, Double)? in
            guard let s = Double(r.score) else { return nil }; return (r, s)
        }.max(by: { $0.1 < $1.1 })?.0
    }

    var worstSubject: ScoreRecord? {
        allScores.compactMap { r -> (ScoreRecord, Double)? in
            guard let s = Double(r.score), s > 0 else { return nil }; return (r, s)
        }.min(by: { $0.1 < $1.1 })?.0
    }

    var averageScore: Double {
        let valid = allScores.compactMap { Double($0.score) }
        guard !valid.isEmpty else { return 0 }
        return valid.reduce(0, +) / Double(valid.count)
    }
}
